Fix crash in rendering on proj 6 builds

nyalldawson opened the pull request:
Ensure PJ* objects are never reused across threads.

Fixes #33902

nyalldawson committed:
Fix crash in rendering on proj 6 builds

Ensure PJ* objects are never reused across threads.

Fixes #33902

nyalldawson commented:
@rouault can i get your eyes on this one?

nirvn commented:
Oh great, this was on my machine the last in-your-face stability issue with windows builds.

rouault commented on src/core/qgscoordinatereferencesystem_p.h:
```diff
@@ -124,6 +136,9 @@ class QgsCoordinateReferenceSystemPrivate : public QSharedData
 #if PROJ_VERSION_MAJOR>=6
+
+    // this is the "master" proj object, to be used as a template for new proj objects created on different threads ONLY.
+    // Always use threadLocalProjObject() instead of this.
```
are we sure that mPj will be destroyed before the PJ_CONTEXT* that owns it ?

nyalldawson replied:
This won't have changed in this PR, but I could add mPj to removeObjectsBelongingToCurrentThread for safety?

nyalldawson replied:
@rouault can you get the context which a PJ* belongs to? Does api exist for that?

rouault replied:
> This won't have changed in this PR

agreed

> but I could add mPj to removeObjectsBelongingToCurrentThread for safety?

then QgsCoordinateReferenceSystemPrivate will have a nullptr mPj and be inconsistent, no ? (in the unlikely scenario that a temporary thread would create a CRS and pass it to the main thread)
But, assuming that a QgsCoordinateReferenceSystem object cannot be simultaneously used by several threads at the same time, I'm wondering if instead of that logic of having a per-thread cache of PJ* objects , you couldn't just use proj_assign_context() with the current TLS context ? And at destruction time of mPj, you create a temporary PJ_CONTEXT like done in QgsCoordinateTransformPrivate::freeProj()

rouault replied:
> can you get the context which a PJ* belongs to?

I don't think so. But proj_assign_context() should be cheap enough to do it systematically

nyalldawson replied:
> then QgsCoordinateReferenceSystemPrivate will have a nullptr mPj and be inconsistent, no 

That's fine -- there's already circumstances when this can be nullptr, so it's already guarded against this.

> But, assuming that a QgsCoordinateReferenceSystem object cannot be simultaneously used by several threads at the same time

it can be -- that's the root cause of the issue here.

> And at destruction time of mPj, you create a temporary PJ_CONTEXT like done in QgsCoordinateTransformPrivate::freeProj()

Will do - incoming shortly.

nyalldawson replied:
> I don't think so. 

No worries - I can easily workaround this by storing the owner PJ_CONTEXT in QgsCoordinateReferenceSystemPrivate

rouault commented on src/core/qgscoordinatereferencesystem_p.h:
```diff
@@ -138,6 +153,49 @@ class QgsCoordinateReferenceSystemPrivate : public QSharedData
 
+#if PROJ_VERSION_MAJOR>=6
+    mutable QReadWriteLock mProjLock;
+    mutable QMap < uintptr_t, PJ * > mProjObjects;
```
any reason to use uintptr_t instead of PJ_CONTEXT* ?

nyalldawson replied:
Nope, good point!

nyalldawson committed:
Use PJ_CONTEXT* instead of void*

jef-n commented:
Guess it'll take to morning in my tz for this to land in 3.10…

nyalldawson committed:
Use a temporary PJ_CONTEXT when destroying PJ objects, just in case the owner context has since been deleted

nyalldawson committed:
Ensure master PJ object is always cleaned up when the context is destroyed too

nyalldawson committed:
Validation hint shouldn't be shared

nyalldawson committed:
Avoid unnecessary duplicate detachments in QgsCoordinateReferenceSystem

In all these code paths a detach had already just occurrred

nyalldawson committed:
Protect some internal members and require access through safe methods

nyalldawson committed:
Don't clone PJ when detaching QgsCoordinateReferenceSystemPrivate

Because in all cases when we are detaching it is because the CRS is
changing, and we will be creating a new PJ anyway

nyalldawson commented:
@jef-n I'm nearly done... about to push a new version (still need to backport though)

nyalldawson committed:
Update expected test result

nyalldawson committed:
Fix test

nyalldawson committed:
Fix equality check for CRS objects which don't have an auth id

nyalldawson committed:
Update test result for proj 4

nyalldawson commented:
Should be good for merging...

nirvn commented:
I seen green everywhere :wink:
